// rclnodejs/generated/dynamixel_sdk_custom_interfaces/dynamixel_sdk_custom_interfaces__msg__SetPosition.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetPositionRefStruct = StructType({
  id: primitiveTypes.uint8,
  position: primitiveTypes.int32,
});
const SetPositionRefArray = ArrayType(SetPositionRefStruct);
const SetPositionRefStructArray = StructType({
  data: SetPositionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetPositionWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetPositionRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._positionIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetPositionRefStruct();
    this._idIntialized = false;
    this._positionIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetPositionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetPositionArrayWrapper;
  }
  static get ArrayType() {
    return SetPositionArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetPositionRefStructArray
  }
  static get refObjectType() {
    return SetPositionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in SetPosition');
      }
      if (checkConsistency && !this._positionIntialized) {
        throw new TypeError('Invalid argument: position in SetPosition');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._positionIntialized = true;
    this._refObject.position = refObject.position;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SetPositionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dynamixel_sdk_custom_interfaces', subFolder: 'msg', interfaceName: 'SetPosition'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get position() {
    if (!this._positionIntialized) {
      return undefined;
    }
    return this._refObject.position;
  }
  set position(value) {
    this._positionIntialized = true;
    this._refObject.position = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetPositionRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._positionIntialized = true;
  }
  copy(other) {
    this._refObject = new SetPositionRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._positionIntialized = true;
  }
  static get classType() {
    return SetPositionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"dynamixel_sdk_custom_interfaces","type":"SetPosition","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetPosition"};
  }
  hasMember(name) {
    let memberNames = ["id","position"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetPositionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetPositionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetPositionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPositionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPositionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetPositionRefArray(size);
    this._refObject = new SetPositionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetPositionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetPositionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetPositionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetPositionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetPositionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetPositionArrayWrapper;
  }
}
module.exports = SetPositionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dynamixel_sdk_custom_interfaces",
    "type": "SetPosition",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetPosition"
}
*/
